### PR TITLE
initial commit of XDR compiler

### PR DESCRIPTION
**Pull request by rlane**
Reviewer: @Sovietaced 

We'll use this program to generate serialization code for IVS's Lua pipeline. Only the Python backend exists at this point. We plan on adding Lua and Java backends too.

### CODE REVIEW

**Comment by Sovietaced** [comment]
Can you explain to me what the optional type is?

File: xdr/parser.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
+# This grammar is written using Pyparsing and closely follows the grammar in
+# RFC 4506.
+#
+# Known differences with the standard:
+#  - We don't allow inline struct, enum, and union definitions.
+#  - The standard is inconsistent on whether identifiers are allowed as enum
+#    values. We only support constants as enum values.
```
Comment (Sovietaced) [inline xdr/parser.py:7]: :+1:

File: xdr/parser.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    kw("struct") - identifier + g(struct_body) + s(";") | \
+    kw("union") - identifier + union_body + s(";")
+
+definition = type_def | constant_def
+
+specification = P.ZeroOrMore(g(definition))
+
+specification.ignore(P.cStyleComment)
+
+def parse_ast(src):
```
Comment (Sovietaced) [inline xdr/parser.py:73]: So parse_ast relies heavily on the correct formatting of the input files (with newlines)?

**Comment by Sovietaced** [comment]
Generally I'd like more comments in your code, so it feels like like a Rich silo of knowledge :)

File: xdr/parser.py (hunk 2 above)
Comment (rlane) [inline xdr/parser.py:73]: Nope, whitespace is ignored by the parser. Like in C some syntax is delimited by semicolons.

**Comment by rlane** [comment]
Thanks for the review, I pushed a couple of commits adding comments to the uglier parts.

**Comment by rlane** [comment]
The optional type is defined at http://tools.ietf.org/html/rfc4506#section-4.19. It's a field that might or might not be present on the wire. It's encoded as a bool followed by the optional field iff the bool is true. In the Python backend we use None to represent a missing optional.

**Pull request merged**